### ISSUE TITLE
nested array index does not work in kernel function

### ISSUE THREAD

**Issue by sroboubi** [gpujs/gpu.js#300]
var gpu = new GPU();

        var broken = this.gpu.createKernel(function(input, lookup)
        {
            return lookup[input[this.thread.x]];
        }).setOutput([1]);

        var working = this.gpu.createKernel(function(input, lookup)
        {
            var idx = input[this.thread.x];
            return lookup[idx];
        }).setOutput([1]);

        console.log(broken([2], [7, 13, 19, 23]));
        console.log(working([2], [7, 13, 19, 23]));

I would expect both kernels to return 19, but the first one (broken) returns 2. Is this a bug?

**Comment by nickhingston** [comment]
this is not a bug.

you can't pass a function to the kernel as a parameter - this is not a 'nested' function.

The function needs to be compiled to shader language at 'compile time', you are passing this function in at runtime.

See createKernelMap documentation - i think this is what you are after...

**Comment by robertleeplummerjr** [comment]
The differences in syntax of

`return lookup[input[this.thread.x]];`

and

```
var idx = input[this.thread.x];
return lookup[idx];
```

shouldn't cause there to be two different results.  I still see this is as a bug.

**Comment by nickhingston** [comment]
i think i added this comment to the wrong issue somehow!

**Comment by robertleeplummerjr** [comment]
In either case, it is a pleasure to work with another human @nickhingston ;)

**Comment by fuzzie360** [comment]
Closing the loop on this seven years later: it was a real bug, and it is now fixed in `develop` (02b346c).

**What was happening.** `lookup[input[this.thread.x]]` compiles to a texture read nested inside another texture read's argument list — both through the same generated helper that takes a `sampler2D` parameter:

```glsl
getMemoryOptimized32(user_lookup, ..., int(getMemoryOptimized32(user_input, ..., threadId.x)))
```

That shape miscompiles in ANGLE's D3D11 backend, which every Chrome, Edge and Firefox on Windows uses for WebGL. HLSL cannot pass samplers as function parameters, so ANGLE specializes sampler-taking functions per sampler — and when a call to the function is nested inside a call to *the same function* with a different sampler, the outer call samples the wrong texture. A raw-WebGL repro with no gpu.js involved shows it directly: `readTex(texB, readTex(texA, 0.5))` reads **texA** twice on D3D11, while a temporary in between, a distinct-but-identical function, or the `texture2D` builtin nested directly are all correct. macOS (ANGLE Metal) and SwiftShader are unaffected, which is why this was so platform-dependent and why the workaround in the original report — assigning to a variable first — worked.

**The fix.** The compiler now hoists a texture read used inside an index expression into a local variable ahead of the statement — the shape that works everywhere. Kernels without the pattern generate byte-identical GLSL to before.

This also turned out to be the cause of a second long-standing failure family: `array[a[x] / b[x]]`-style kernels ("integer division as index") nest reads the same way. With the fix, the full browser suite passes 3126/3126 on Chrome and Edge on Windows 11, where both families previously failed.

An upstream report against ANGLE is being prepared with the standalone repro.
